### ISSUE TITLE
newSuspendedTransaction doesn't work after failure

### ISSUE THREAD

**Issue by davydes** [JetBrains/Exposed#1075]
Example:

```kotlin
runBlocking {
  try {
    newSuspendedTransaction {
      println("Try to insert...")
      throw RuntimeException("Fail")
    }
  } catch (i: Exception) {
    println("Seems record exists, then...")
  }

  newSuspendedTransaction {
    println("Try to update...")
    println("Success")
  }
}
```

Produces output:
```
Try to insert...
Seems record exists, then...
```

**Comment by Tapac** [comment]
Hi @davydes , it's a normal behavior as the whole `runBlocking` scope is interrupted when the exception is thrown.
Please read more about exceptions and coroutines [here](https://kotlinlang.org/docs/reference/coroutines/exception-handling.html) or [here](https://medium.com/androiddevelopers/exceptions-in-coroutines-ce8da1ec060c)

**Comment by davydes** [comment]
> Hi @davydes , it's a normal behavior as the whole `runBlocking` scope is interrupted when the exception is thrown.
> Please read more about exceptions and coroutines [here](https://kotlinlang.org/docs/reference/coroutines/exception-handling.html) or [here](https://medium.com/androiddevelopers/exceptions-in-coroutines-ce8da1ec060c)

But seems it breaks logic, i didn't run any job here, I just run coroutine and handle exception which it throws. Why it interrupts my parent CoroutineScope? In links you provided explanation shows only case when i run launch with unhandled exception inside.

**Comment by davydes** [comment]
Also I can rewrite code:

```kotlin
runBlocking {
  try {
    withContext(IO) {
      transaction {
        println("Try to insert...")
        throw RuntimeException("Fail")
      }
    }
  } catch (i: Exception) {
    println("Seems record exists, then...")
  }

  withContext(IO) {
    transaction {
      println("Updating...")
      println("Success!")
    }
  }
}
```

And it will produce next output:

```
Try to insert...
Seems record exists, then...
Updating...
Success!
```

Why newSuspendedTransaction works in different way?

**Comment by Tapac** [comment]
The same here  with `newSuspendedTransaction` if you don't provide a specific context it will be executed within current and when failed the whole context execution will be prevented.
You could provide the context as a first parameter like `newSuspendedTransaction(IO) {}` then your code will run as expected

**Comment by nomisRev** [comment]
> Hi @davydes , it's a normal behavior as the whole runBlocking scope is interrupted when the exception is thrown.
Please read more about exceptions and coroutines [here](https://kotlinlang.org/docs/reference/coroutines/exception-handling.html) or [here](https://medium.com/androiddevelopers/exceptions-in-coroutines-ce8da1ec060c)

@Tapac,If I'm not mistaken this breaks the semantics of Structured Concurrency as it's followed by Kotlinx Coroutines. If `newSuspendedTransaction` _cancels_ the `parent`, then it should also throw `CancellationException`. 

Looking at the code, https://github.com/JetBrains/Exposed/blob/990bff364176716e92cc1aac3a0217940da96513/exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/transactions/experimental/Suspended.kt#L21 `TransactionScope` creates a new `CoroutineScope` with the `runBlocking` as parent. Which is effectively the same as `withContext` or `coroutineScope`, but these builders behave entirely differently.

```kotlin
runBlocking {
  kotlin.runCatching {
    coroutineScope {
      println("1")
      throw IllegalStateException("Boom!")
    }
  }.let(::println)

  coroutineScope {
    println("2")
  }
}
```
```text
1
Failure(java.lang.IllegalStateException: Boom!)
2
```

**Comment by j30ng** [comment]
@nomisRev To be fair, this behavior rather complies with the semantics of structured concurrency, because it cancels the parent in case of an unhandled exception.

Try opening supervisorScope around newSuspendedTransaction, like:
```kotlin
try {
    supervisorScope {
        newSuspendedTransaction {
            // ...
        }
    }
}
```

**Comment by nomisRev** [comment]
@j30ng thank you for the example. To rephrase then, and I think I'm still in line with the original bug report.

Semantically it still doesn't seem sense to me, because `newSuspendedTransaction` has the same signature as `coroutineScope` or `withContext` so I expect it to behave the same. Further more, I do not expect that I have to wrap it into `supervisorScope` to run it if I want to catch any exceptions.

If not, I'd expect it to actually work like `async` which is of shape `fun CoroutineScope.async(...): Deferred<T>`. The reason that `async` works the way you describe is because it's not a `suspend` function and is thus _unscoped_ through `suspend` unlike `coroutineScope` or `newSuspendedTransaction`.

This can be shown by commenting `await` in your example, meaning that the exceptions *has to* be caught inside `async` itself if you want `2` to be printed.

```kotlin
fun main() = runBlocking {
    coroutineScope {
        kotlin.runCatching {
            async {
                println("1")
                throw IllegalStateException("Boom!")
            }//.await()
        }.let(::println)
    }

    println("2")
}
```

Still prints

```kotlin
Success(DeferredCoroutine{Active}@63e31ee)
1
Exception in thread "main" java.lang.IllegalStateException: Boom!
	at my.timer.TreeKt$main$1$1$1$1.invokeSuspend(tree.kt:42)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:280)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	at my.timer.TreeKt.main(tree.kt:37)
	at my.timer.TreeKt.main(tree.kt)
```

So to recap my thoughts:

`suspend fun <A> newSuspendedTransaction(..): A` should in my opinion behave like `suspend fun <A> withContext(...): A` or `suspend fun <A> coroutineScope(..): A`.

If it's desired to behave like `fun <A> CoroutineScope.async(...): Deferred<A>` or `fun <A> launch(...): A` I think it should match the same signature, `fun <A> CoroutineScope.newSuspendedTransaction(...): Deferred<A>,` but I think that is much less desirable. Otherwise I feel like we're comparing oranges to apples if we are not comparing functions with the same signatures.

**Comment by matejdro** [comment]
Agree here, if I already `catch` the exception, then that exception should not propagate to the parent scope. This breaks usual coroutine behavior.

Here is a more fleshed version of above workaround that we are using in our code:

```kotlin
/**
 * Wrapper for [newSuspendedTransaction] that fixes error propagation.
 * A workaround for https://github.com/JetBrains/Exposed/issues/1075.
 */
suspend fun <T> newSuspendedTransactionWithExceptionHandling(
   context: CoroutineContext? = null,
   db: Database? = null,
   transactionIsolation: Int? = null,
   statement: suspend Transaction.() -> T
): T {
   return supervisorScope {
      newSuspendedTransaction(context, db, transactionIsolation, statement)
   }
}
```

**Comment by obabichevjb** [comment]
This issue should be fixed in version 1.0.0-rc-1 through PR [#2601](https://github.com/JetBrains/Exposed/pull/2601).

  The problem was resolved by refactoring of `withTransactionScope()`, so `withContext()` is used now. Caught exceptions should not cancel the parent coroutine scope anymore.

  Note that `newSuspendedTransaction` is now deprecated. Please migrate to `suspendTransaction()`. If you need a specific context, wrap it with `withContext()`
